File: kpis/kpiYawRate.m
function kpiYawRate(obj, i, aebStartIdx, yawRateSuspTh)
    % KPI Yaw Rate Analysis
    % This function analyzes the yaw rate signal to determine the maximum yaw rate
    % and whether it exceeds a specified threshold after an AEB request event.
    %% Inputs:
    % obj            - KPIExtractor object
    % i              - Current index in the kpiTable
    % aebStartIdx    - Index of AEB request event
    % yawRateSuspTh  - Threshold for yaw rate suspension in degrees per second

    % Use obj.kpiTable and obj.signalMatChunk directly
    kpiTable = obj.kpiTable;
    signalMatChunk = obj.signalMatChunk;
    offset = obj.TIME_IDX_OFFSET;

    % Ensure required columns exist
    if ~ismember('yawRateMax', kpiTable.Properties.VariableNames)
        obj.kpiTable.yawRateMax = NaN(height(obj.kpiTable), 1);
    end
    if ~ismember('absYawRateMaxDeg', kpiTable.Properties.VariableNames)
        obj.kpiTable.absYawRateMaxDeg = NaN(height(obj.kpiTable), 1);
    end
    if ~ismember('isYawRateHigh', kpiTable.Properties.VariableNames)
        obj.kpiTable.isYawRateHigh = false(height(obj.kpiTable), 1);
    end

    % Calculate yaw rate KPIs
    [~, idx] = max(abs(signalMatChunk.yawRate(aebStartIdx - offset:end)));
    yawRateMax = signalMatChunk.yawRate(aebStartIdx - offset + idx - 1);
    absYawRateMaxDeg = round(abs(yawRateMax * 180 / pi()), 2);

    obj.kpiTable.yawRateMax(i) = yawRateMax;
    obj.kpiTable.absYawRateMaxDeg(i) = absYawRateMaxDeg;
    obj.kpiTable.isYawRateHigh(i) = absYawRateMaxDeg > yawRateSuspTh;
end

File: python/kpis/aeb/kpi_yaw_rate.py
import numpy as np
import pandas as pd

def kpi_yaw_rate(obj, i, aeb_start_idx, yaw_rate_susp_th):
    """
    KPI Yaw Rate Analysis during an AEB event.
    Updates obj.kpi_table in place.

    Parameters
    ----------
    obj : object with attributes
        - kpi_table : pandas.DataFrame
        - signal_mat_chunk : dict with array "yawRate" (rad/s)
        - TIME_IDX_OFFSET : int
    i : int
        Row index in kpi_table to update.
    aeb_start_idx : int
        Index of AEB request event.
    yaw_rate_susp_th : float
        Threshold for yaw rate suspension in degrees/sec.
    """
    kpi_table = obj.kpi_table
    signals = obj.signal_mat_chunk
    offset = obj.TIME_IDX_OFFSET

    # Ensure required columns exist
    for col, default, dtype in [
        ("yawRateMax", np.nan, "float"),
        ("absYawRateMaxDeg", np.nan, "float"),
        ("isYawRateHigh", False, "bool"),
    ]:
        if col not in kpi_table.columns:
            kpi_table[col] = pd.Series([default] * len(kpi_table), dtype=dtype)

    # Slice yaw rate signal
    segment = signals["yawRate"][aeb_start_idx - offset :]
    idx_rel = int(np.argmax(np.abs(segment)))
    yaw_rate_max = segment[idx_rel]
    abs_yaw_rate_max_deg = round(abs(yaw_rate_max * 180 / np.pi), 2)

    # Update table
    kpi_table.at[i, "yawRateMax"] = float(yaw_rate_max)
    kpi_table.at[i, "absYawRateMaxDeg"] = abs_yaw_rate_max_deg
    kpi_table.at[i, "isYawRateHigh"] = abs_yaw_rate_max_deg > yaw_rate_susp_th
